README
LABADMIN
--------

This is a program that enables easy administration of multiple gnu/linux systems.




debian/changelog
labadmin (0.1-1.2) unstable; urgency=low

   * Adapted for Nmap 5.21 

 -- pnegre <pere.negre@gmail.com>  Fri, 02 Sep 2011 12:57:53 +0200

labadmin (0.1-1.1) unstable; urgency=low

  * Changed dependencies

 -- pnegre <pnegre@andromeda.lan>  Fri, 30 Apr 2010 20:08:12 +0200

labadmin (0.1-1) unstable; urgency=low

  * First release.

 -- pnegre <pere.negre@gmail.com>  Wed, 28 Apr 2010 22:10:50 +0200


stuff/lab_admin.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys, os, re, time
from PyQt4 import QtCore, QtGui, uic

import utils


class HostWgt(QtGui.QTableWidgetItem):
	def __init__(self,p,t):
		QtGui.QTableWidgetItem.__init__(self,t)
		self.taskObject = p
		self.setFlags(QtCore.Qt.ItemIsSelectable | QtCore.Qt.ItemIsEnabled)




class HostItem(object):
	def setup(self,ip,mac):
		self.ip = ip
		self.mac = mac
		self.tag = None
	
	
	def insert(self,t):
		t.insertRow(0)
		t.setItem(0, 0, HostWgt(self,str(self.ip)))
		t.setItem(0, 1, HostWgt(self,str(self.mac)))
		t.setItem(0, 2, HostWgt(self,str(self.tag)))


class PBarDlg(QtGui.QDialog):
	def __init__(self, parent=None):
		QtGui.QWidget.__init__(self,parent)
		self.ui = uic.loadUi("pbar.ui",self)



class ClusterDlg(QtGui.QDialog):
	def __init__(self, parent=None):
		QtGui.QWidget.__init__(self,parent)
		self.ui = uic.loadUi("cluster.ui",self)



def get_macs(hosts):
	p = QtCore.QProcess()
	p.start(utils.completePath("arp"), ["-n"])
	p.waitForFinished(-1)
	dta = p.readAll()
	macs = {}
	for h in hosts:
		macs[h] = None
		m = re.search(h + '\s+ether\s+(\S+)', str(dta))
		if m:
			macs[h] = m.group(1).lower()
	return macs


def search_hosts(network,win):
	p = QtCore.QProcess()
	pBar = PBarDlg(win)
	pBar.show()
	p.start(utils.completePath("nmap"), ["-n", "-sP", "-oG", "-", network])
	while not p.waitForFinished(30):
		QtGui.QApplication.processEvents()
	dta = p.readAll()
	m = re.findall('Host: (\S+) .* Up',str(dta))
	pBar.close()
	return m




class Filter(object):
	def __init__(self):
		self.clear()
	
	
	def loadFromFile(self, file):
		self.clear()
		try:
			f = open(file,"r")
		except(IOError), e:
			print "error opening", file
			return False
		lines = f.readlines()
		f.close()
		mac = ""; tag=""
		for l in lines:
			l = l.rstrip("\n")
			m = re.search("^(..:..:..:..:..:..)\s+(\S*)$",l)
			if m:
				mac = m.group(1)
				tag = m.group(2)
			else:
				m = re.search("^(..:..:..:..:..:..)\s*$",l)
				if not m: continue
				mac = m.group(1)
				tag = None
			mac = mac.lower()
			self.macs.append(mac)
			self.tags[mac] = tag
		return True
	
	
	def clear(self):
		self.macs = []
		self.tags = {}

	
	def exe(self, hostList = []):
		r = []
		for h in hostList:
			if h.mac == None: continue
			if h.mac in self.macs:
				h.tag = self.tags[h.mac]
				r.append(h)
		return r


class MainWindow(QtGui.QMainWindow):
	def __init__(self, parent=None):
		QtGui.QWidget.__init__(self,parent)
		self.ui = uic.loadUi("mainwindow.ui",self)
		self.ui.hostList.horizontalHeader().setStretchLastSection(True)
		self.hList = []
		self.filteredList = self.hList
		self.fnameList = []
		
		self.setWindowTitle("Lab Admin")

		self.connect(self.ui.buttonHosts,
			QtCore.SIGNAL("clicked()"), self.getHosts)
		self.connect(self.ui.buttonFilter,
			QtCore.SIGNAL("clicked()"), self.loadFilter)
		self.connect(self.ui.buttonClearFilters,
			QtCore.SIGNAL("clicked()"), self.clearFilters)
		self.connect(self.ui.buttonCluster,
			QtCore.SIGNAL("clicked()"), self.execCluster)
		self.connect(self.ui.filterBox,
			QtCore.SIGNAL("activated(int)"), self.applyFilter)
		
		self.center()
		self.ui.buttonCluster.setDisabled(True)
		self.ui.hostList.setColumnWidth(0,150)
		self.ui.hostList.setColumnWidth(1,150)
		
		self.settings = QtCore.QSettings("labadmin","labadmin")
		self.loadSettings()
		
		
	def loadSettings(self):
		s = self.settings.value("filters").toString()
		if s == "": return
		#print "-" + s + "-"
		l = str(s).split("|")
		
		for f in l:
			#print f
			self.loadFilter(f)
		
	
	def center(self):
		screen = QtGui.QDesktopWidget().screenGeometry()
		size =  self.geometry()
		self.move((screen.width()-size.width())/2, (screen.height()-size.height())/2)
	
	
	
	def closeEvent(self,e):
		v = "|".join(self.fnameList)
		self.settings.setValue("filters", QtCore.QVariant(v))
		e.accept()
	
	
	def execCluster(self):
		d = ClusterDlg()
		if not d.exec_(): return
		us = d.username.text()
		tl = "-G"
		if d.tileWindows.isChecked(): tl = "-g"
		hosts = []
		for h in self.filteredList:
			hosts.append(str(h.ip))
		pid = os.fork()
		if pid == 0:
			os.execl(utils.completePath("cssh"), "cssh", tl, "-l" + us, *hosts)		
	
	
	def clearTable(self):
		while self.ui.hostList.item(0,0):
			self.ui.hostList.removeRow(0)
	
	
	def refreshTable(self):
		self.clearTable()
		for h in self.filteredList:
			h.insert(self.ui.hostList)
		
		if len(self.filteredList):
			self.ui.buttonCluster.setDisabled(False)
		else:
			self.ui.buttonCluster.setDisabled(True)
		
		
	def getHosts(self):
		n,c = QtGui.QInputDialog.getText(self,"Network","What network? (ex: 192.168.2.0/24)")
		if not c: return		
		hosts = search_hosts(n,self)
		macs = get_macs(hosts)
		self.hList = []
		for h in hosts:
			hi = HostItem()
			hi.setup(h,macs[h])
			self.hList.append(hi)
		
		self.filteredList = self.hList
		self.refreshTable()
	
	
	def loadFilter(self, fnd=''):
		fn = ''
		if fnd == '':
			fn = QtGui.QFileDialog.getOpenFileName(self, "Load File")
			if fn == '': return
		else:
			fn = fnd
		if fn in self.fnameList: return
		f = Filter()
		f.clear()
		if not f.loadFromFile(fn): return
		self.ui.filterBox.addItem("..." + fn[-20:], QtCore.QVariant(f))
		self.fnameList.append(str(fn))


	def applyFilter(self,i):
		if i==0:
			self.filteredList = self.hList
		else:
			f = self.ui.filterBox.itemData(i).toPyObject()
			self.filteredList = f.exe(self.hList)
		self.refreshTable()

	def clearFilters(self):
		self.fnameList = []
		self.ui.filterBox.clear()
		self.ui.filterBox.addItem('All hosts')



def main():
	if utils.checkRequiredPrograms() == False:
		exit(0)
	
	app = QtGui.QApplication(sys.argv)
	mainWin = MainWindow()
	mainWin.show()
	app.exec_()


labadmin.py
#!/usr/bin/python
# -*- coding: utf-8 -*-

import sys, os, re, time
from PyQt4 import QtCore, QtGui, uic

import utils


class HostWgt(QtGui.QTableWidgetItem):
	def __init__(self,p,t):
		QtGui.QTableWidgetItem.__init__(self,t)
		self.taskObject = p
		self.setFlags(QtCore.Qt.ItemIsSelectable | QtCore.Qt.ItemIsEnabled)




class HostItem(object):
	def setup(self,ip,mac):
		self.ip = ip
		self.mac = mac
		self.tag = None
	
	
	def insert(self,t):
		t.insertRow(0)
		t.setItem(0, 0, HostWgt(self,str(self.ip)))
		t.setItem(0, 1, HostWgt(self,str(self.mac)))
		t.setItem(0, 2, HostWgt(self,str(self.tag)))


class PBarDlg(QtGui.QDialog):
	def __init__(self, parent=None):
		QtGui.QWidget.__init__(self,parent)
		self.ui = uic.loadUi("pbar.ui",self)



class ClusterDlg(QtGui.QDialog):
	def __init__(self, parent=None):
		QtGui.QWidget.__init__(self,parent)
		self.ui = uic.loadUi("cluster.ui",self)



def get_macs(hosts):
	p = QtCore.QProcess()
	p.start(utils.completePath("arp"), ["-n"])
	p.waitForFinished(-1)
	dta = p.readAll()
	macs = {}
	for h in hosts:
		macs[h] = None
		m = re.search(h + '\s+ether\s+(\S+)', str(dta))
		if m:
			macs[h] = m.group(1).lower()
	return macs


def search_hosts(network,win):
	p = QtCore.QProcess()
	pBar = PBarDlg(win)
	pBar.show()
	p.start(utils.completePath("nmap"), ["-n", "-sP", network])
	while not p.waitForFinished(30):
		QtGui.QApplication.processEvents()
	dta = p.readAll()
	m = re.findall('Host (\S+) .* up',str(dta))
	pBar.close()
	return m




class Filter(object):
	def __init__(self):
		self.clear()
	
	
	def loadFromFile(self, file):
		self.clear()
		try:
			f = open(file,"r")
		except(IOError), e:
			print "error opening", file
			return False
		lines = f.readlines()
		f.close()
		mac = ""; tag=""
		for l in lines:
			l = l.rstrip("\n")
			m = re.search("^(..:..:..:..:..:..)\s+(\S*)$",l)
			if m:
				mac = m.group(1)
				tag = m.group(2)
			else:
				m = re.search("^(..:..:..:..:..:..)\s*$",l)
				if not m: continue
				mac = m.group(1)
				tag = None
			mac = mac.lower()
			self.macs.append(mac)
			self.tags[mac] = tag
		return True
	
	
	def clear(self):
		self.macs = []
		self.tags = {}

	
	def exe(self, hostList = []):
		r = []
		for h in hostList:
			if h.mac == None: continue
			if h.mac in self.macs:
				h.tag = self.tags[h.mac]
				r.append(h)
		return r


class MainWindow(QtGui.QMainWindow):
	def __init__(self, parent=None):
		QtGui.QWidget.__init__(self,parent)
		self.ui = uic.loadUi("mainwindow.ui",self)
		self.ui.hostList.horizontalHeader().setStretchLastSection(True)
		self.hList = []
		self.filteredList = self.hList
		self.fnameList = []
		
		self.setWindowTitle("Lab Admin")

		self.connect(self.ui.buttonHosts,
			QtCore.SIGNAL("clicked()"), self.getHosts)
		self.connect(self.ui.buttonFilter,
			QtCore.SIGNAL("clicked()"), self.loadFilter)
		self.connect(self.ui.buttonClearFilters,
			QtCore.SIGNAL("clicked()"), self.clearFilters)
		self.connect(self.ui.buttonCluster,
			QtCore.SIGNAL("clicked()"), self.execCluster)
		self.connect(self.ui.filterBox,
			QtCore.SIGNAL("activated(int)"), self.applyFilter)
		
		self.center()
		self.ui.buttonCluster.setDisabled(True)
		self.ui.hostList.setColumnWidth(0,150)
		self.ui.hostList.setColumnWidth(1,150)
		
		self.settings = QtCore.QSettings("labadmin","labadmin")
		self.loadSettings()
		
		
	def loadSettings(self):
		s = self.settings.value("filters").toString()
		if s == "": return
		#print "-" + s + "-"
		l = str(s).split("|")
		
		for f in l:
			#print f
			self.loadFilter(f)
		
	
	def center(self):
		screen = QtGui.QDesktopWidget().screenGeometry()
		size =  self.geometry()
		self.move((screen.width()-size.width())/2, (screen.height()-size.height())/2)
	
	
	
	def closeEvent(self,e):
		v = "|".join(self.fnameList)
		self.settings.setValue("filters", QtCore.QVariant(v))
		e.accept()
	
	
	def execCluster(self):
		d = ClusterDlg()
		if not d.exec_(): return
		us = d.username.text()
		tl = "-G"
		if d.tileWindows.isChecked(): tl = "-g"
		hosts = []
		for h in self.filteredList:
			hosts.append(str(h.ip))
		pid = os.fork()
		if pid == 0:
			os.execl(utils.completePath("cssh"), "cssh", tl, "-l" + us, *hosts)		
	
	
	def clearTable(self):
		while self.ui.hostList.item(0,0):
			self.ui.hostList.removeRow(0)
	
	
	def refreshTable(self):
		self.clearTable()
		for h in self.filteredList:
			h.insert(self.ui.hostList)
		
		if len(self.filteredList):
			self.ui.buttonCluster.setDisabled(False)
		else:
			self.ui.buttonCluster.setDisabled(True)
		
		
	def getHosts(self):
		n,c = QtGui.QInputDialog.getText(self,"Network","What network? (ex: 192.168.2.0/24)")
		if not c: return		
		hosts = search_hosts(n,self)
		macs = get_macs(hosts)
		self.hList = []
		for h in hosts:
			hi = HostItem()
			hi.setup(h,macs[h])
			self.hList.append(hi)
		
		self.filteredList = self.hList
		self.refreshTable()
	
	
	def loadFilter(self, fnd=''):
		fn = ''
		if fnd == '':
			fn = QtGui.QFileDialog.getOpenFileName(self, "Load File")
			if fn == '': return
		else:
			fn = fnd
		if fn in self.fnameList: return
		f = Filter()
		f.clear()
		if not f.loadFromFile(fn): return
		self.ui.filterBox.addItem("..." + fn[-20:], QtCore.QVariant(f))
		self.fnameList.append(str(fn))


	def applyFilter(self,i):
		if i==0:
			self.filteredList = self.hList
		else:
			f = self.ui.filterBox.itemData(i).toPyObject()
			self.filteredList = f.exe(self.hList)
		self.refreshTable()

	def clearFilters(self):
		self.fnameList = []
		self.ui.filterBox.clear()
		self.ui.filterBox.addItem('All hosts')



if __name__ == '__main__':
	if utils.checkRequiredPrograms() == False:
		exit(0)
	
	app = QtGui.QApplication(sys.argv)
	mainWin = MainWindow()
	mainWin.show()
	app.exec_()
